wrappers/redis/redis.go

```diff
@@ -0,0 +1,169 @@
+package epsagonredis
+
+import (
+	"context"
+	"encoding/json"
+	"net"
+	"runtime/debug"
+
+	"github.com/epsagon/epsagon-go/epsagon"
+	"github.com/epsagon/epsagon-go/protocol"
+	"github.com/epsagon/epsagon-go/tracer"
+	"github.com/go-redis/redis/v8"
+	"github.com/google/uuid"
+)
+
+type epsagonHook struct {
+	host    string
+	port    string
+	dbIndex string
+	event   *protocol.Event
+}
+
+func NewClient(opt *redis.Options) *redis.Client {
+	client := redis.NewClient(opt)
+	return wrapClient(client, opt)
+}
+
+func wrapClient(client *redis.Client, opt *redis.Options) *redis.Client {
+	host, port := getClientHostPort(opt)
+	client.AddHook(&epsagonHook{
+		host:    host,
+		port:    port,
+		dbIndex: string(opt.DB),
+	})
+	return client
+}
+
+func getClientHostPort(opt *redis.Options) (host, port string) {
+	if opt.Network == "unix" {
+		return "localhost", opt.Addr
+	}
+	host, port, err := net.SplitHostPort(opt.Addr)
+	if err == nil {
+		return host, port
+	}
+	return opt.Addr, ""
+}
+
+func (epsHook *epsagonHook) BeforeProcess(ctx context.Context, cmd redis.Cmder) (context.Context, error) {
+	cmdArgs := safeJsonify(cmd.Args())
+	return epsHook.before(ctx, cmd.Name(), cmdArgs)
+}
+
+func (epsHook *epsagonHook) AfterProcess(ctx context.Context, cmd redis.Cmder) error {
+	var errMsg string
+	if err := cmd.Err(); err != nil {
+		errMsg = err.Error()
+	}
+	return epsHook.after(ctx, cmd.String(), errMsg)
+}
+
+func (epsHook *epsagonHook) BeforeProcessPipeline(ctx context.Context, cmds []redis.Cmder) (context.Context, error) {
+	cmdArgs := safeJsonify(getPiplineCmdArgs(cmds))
+	return epsHook.before(ctx, "Pipeline", cmdArgs)
+}
+
+func (epsHook *epsagonHook) AfterProcessPipeline(ctx context.Context, cmds []redis.Cmder) error {
+	var errMsg string
+	if errors := getPipelineErrors(cmds); len(errors) > 0 {
+		errMsg = safeJsonify(errors)
+
+	}
+	response := safeJsonify(getPipelineResponse(cmds))
+	return epsHook.after(ctx, response, errMsg)
+}
+
+func (epsHook *epsagonHook) before(ctx context.Context, operation, cmdArgs string) (context.Context, error) {
+	metadata := getResourceMetadata(epsHook, cmdArgs)
+	epsHook.event = createEvent(epsHook, operation, metadata)
```

Comment:
make sure that in the new fixes, the create event will be done only if there's a tracer